ansible/roles/docker/tasks/main.yml

```diff
@@ -38,29 +46,69 @@
   when: ansible_os_family == "RedHat"
 
 - name: check docker-tcp socket state
-  shell: systemctl status docker-tcp.socket | grep "Active:"
+  shell: systemctl status docker-tcp.socket | grep 'Active.*active' -o
   register: docker_tcp_socket_state
 
+- include: create_docker_device.yml
+  when: docker_device != ""
+
 - name: copy systemd units for docker tcp socket settings
   template: src=docker-tcp.j2 dest=/etc/systemd/system/docker-tcp.socket
   register: docker_tcp_socket
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload and stop docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: 
```

Comment:
Yup

On 29 Feb 2016, at 10:19, Madhav Puri wrote:

>> tags:
>>   - prebake-for-dev
>>
>> +# XXX: this needs to happen twice after setting up the thin 
>> provisioner because
>> +#      of some docker bug I've not investigated.
>> +- name: restart docker (first time)
>> +  service:
>> +    name: docker
>> +    state: restarted
>> +  when:
>
>> explicitly, not unconditionally
>
> ah, I see but note that this is topped only when tcp-socket service is 
> setup for first time i.e. this condition will not happen on subsequent 
> re-run of the playbook as tcp-service would already have been setup.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/contiv/volplugin/pull/152/files#r54452215